fuzzing/private/engine.bzl

```diff
@@ -44,18 +45,33 @@ def _cc_fuzzing_engine_impl(ctx):
         launcher_runfiles = launcher_runfiles.merge(ctx.runfiles(files = data_files))
         launcher_runfiles = launcher_runfiles.merge(data_label[DefaultInfo].default_runfiles)
 
-    cc_library_info = ctx.attr.library[CcInfo]
-    cc_fuzzing_engine_info = CcFuzzingEngineInfo(
+    cc_library_info = None
+    java_library_info = None
+    default_info = None
+    providers = []
+    if ctx.attr.library and not ctx.attr.java_library:
+        cc_library_info = ctx.attr.library[CcInfo]
+        default_info = ctx.attr.library[DefaultInfo]
+        providers += [default_info, cc_library_info]
+    elif ctx.attr.java_library and not ctx.attr.library:
+        java_library_info = ctx.attr.java_library[JavaInfo]
+        default_info = ctx.attr.java_library[DefaultInfo]
+        providers += [default_info, java_library_info]
+    else:
+        fail("A fuzzing_engine must have exactly one of library and java_library set.")
+
+    fuzzing_engine_info = FuzzingEngineInfo(
```

Comment:
Design question: Are we sharing the same provider here in order to reuse the shared logic of populating these fields?

I'm wondering if we could still take advantage of the Bazel provider type system and define a `CommonFuzzingEngineInfo` provider, that can be nested inside separate `CcFuzzingEngineInfo` and `JavaFuzzingEngineInfo` providers.

Reply:
If it's possible to nest providers than this would certainly be the preferable solution. Do you know whether that is possible? The documentation at https://docs.bazel.build/versions/master/skylark/lib/Provider.html suggests it may not be.

Reply:
That should be possible - we are actually already nesting providers this way (e.g., the `FuzzingEngienInfo.java_library_info` field includes a `JavaInfo` provider.

I'm basically thinking of the following setup:

```
CommonFuzzingEngineInfo = provider(
   ...
)

CcFuzzingEngineInfo = provider(
   cc_library_info = ...
   engine_info = (the CommonFuzzingEngineInfo data)
)

JavaFuzzingEngineInfo = provider(
    java_library_info = ...
    engine_info = (CommonFuzzingEngineInfo data)
)
```

We can then have two distinct rules java_fuzzing_engine and cc_fuzzing_engine instantiating these providers. We can factor out shared functionality (the dict of shared attributes, a shared function to compute CommonFuzzingEngineInfo).

WDYT?

Reply:
How about we drop the `*_library_info` field entirely in favor of a single, language-independent `FuzzingEngineInfo` provider? Engines already expose their libraries via a dedicated `{Cc, Java}Info` provider and the single use of that info in the code base is via that provider anyway. 

Composing providers rather than nesting them also interacts better with the `provides` directive of the `engine` attribute in `fuzzing_binary`. So even if we want to keep language-specific engine info providers, it may be easier to split them into language-dependent and language-independent parts rather than nesting them. What do you think?

Reply:
Do you mean to basically say that attributes depending on a fuzzing engine should explicitly list the extra `{Cc,Java}Info` providers they might need? E.g.:

```
"engine": attr.label(providers=[FuzzingEngineInfo,CcInfo]),
```

I think that's a very good idea and a better use of Bazel providers than what we currently do!

Would it make sense to do this first in a separate PR, just so we don't mix semantics-preserving refactorings with adding new functionality?

Reply:
I submitted #144 with just this change. As it turns out, there are no attributes that explicitly depend on the language info provider, which makes the refactoring even easier.